lib/global_uid/base.rb

```diff
@@ -59,22 +59,22 @@ def self.with_servers
         servers.shuffle! # subsequent requests are made against different servers
       end
 
-      raise NoServersAvailableException if servers.empty?
-
       errors = []
       servers.each do |server|
         begin
           yield server if server.active?
         rescue TimeoutException, Exception => e
-          notify e, "#{e.message}"
+          notify(e, e.message)
           errors << e
           server.disconnect!
           server.update_retry_at(1.minute)
         end
       end
 
       if connections.empty?
-        raise NoServersAvailableException, "Errors hit: #{errors.map(&:to_s).join(',')}"
+        exception = NoServersAvailableException.new(errors.empty? ? "" : "Errors hit: #{errors.map(&:to_s).join(', ')}")
+        notify(exception, exception.message)
```

Comment:
This line is new – and not a refactoring.

Reply:
Good catch, that should be a separate commit, we should call `notify` before raising the exception